examples/image_captioning/model.py

```diff
@@ -0,0 +1,283 @@
+import numpy as np
+
+import chainer
+from chainer import functions as F
+from chainer import initializers
+from chainer import links as L
+from chainer import reporter
+from chainer import Variable
+
+
+class ImageCaptionModel(chainer.Chain):
+
+    """Image captioning model."""
+
+    def __init__(self, vocab_size, hidden_size=512, rnn='lstm',
+                 dropout_ratio=0.5, finetune_feat_extractor=False,
+                 ignore_label=-1):
+        super(ImageCaptionModel, self).__init__()
+
+        if rnn == 'lstm':
+            LanguageModel = LSTMLanguageModel
+        elif rnn == 'nsteplstm':
+            LanguageModel = NStepLSTMLanguageModel
+        else:
+            raise ValueError('Invalid RNN type.')
+
+        with self.init_scope():
+            self.feat_extractor = VGG16FeatureExtractor()
+            self.lang_model = LanguageModel(
+                vocab_size, hidden_size, dropout_ratio=dropout_ratio,
+                ignore_label=ignore_label)
+
+        self.finetune_feat_extractor = finetune_feat_extractor
+
+    def prepare(self, img):
+        """Single image to resized and normalized image."""
+        return self.feat_extractor.prepare(img)
+
+    def __call__(self, imgs, captions):
+        """Batch of images to a single loss."""
+        imgs = Variable(imgs)
+        if self.finetune_feat_extractor:
+            img_feats = self.feat_extractor(imgs)
+        else:
+            # Extract features with the `train` configuration set to `False` in
+            # order to basically skip the dropout regularizations. This is how
+            # dropout is used during standard inference. Also, since we are not
+            # going to optimize the feature extractor, we explicitly set the
+            # backpropgation mode to not construct any computational graphs.
+            with chainer.using_config('train', False), \
+                    chainer.no_backprop_mode():
+                img_feats = self.feat_extractor(imgs)
+
+        loss = self.lang_model(img_feats, captions)
+
+        # Report the loss so that it can be printed, logged and plotted by
+        # other trainer extensions
+        reporter.report({'loss': loss}, self)
+
+        return loss
+
+    def predict(self, imgs, bos, eos, max_caption_length):
+        """Batch of images to captions."""
+        imgs = Variable(imgs)
+        img_feats = self.feat_extractor(imgs)
+        captions = self.lang_model.predict(
+            img_feats, bos=bos, eos=eos, max_caption_length=max_caption_length)
+        return captions
+
+
+class VGG16FeatureExtractor(chainer.Chain):
+
+    """VGG16 image feature extractor."""
+
+    def __init__(self):
+        super(VGG16FeatureExtractor, self).__init__()
+        with self.init_scope():
+            self.cnn = L.VGG16Layers()
+        self.cnn_layer_name = 'fc7'
+
+    def prepare(self, img):
+        """Single image to resized and normalized image."""
```

Comment:
Could you explain that RGB-to-BGR conversion is performed here?
It is hard for beginners to guess what `L.model.vision.vgg.prepare` does at a glance.